### PR TITLE
[FlagGems Operator Development Competition] median

### PR DESCRIPTION
**Pull request by gaibianshiji**
## Overview

Implement `median.dim` operator using Triton, computing median along a specified dimension and returning both values and indices tensors.

**Operator Schema**:
```
median.dim(Tensor self, int dim, bool keepdim=False) -> (Tensor values, Tensor indices)
```

### Algorithm

Full sort + median extraction approach:

1. Load input rows with power-of-2 padding (using max float value to avoid `tl.sort` producing `nan` with `inf`)
2. Replace `inf`/`-inf` with large finite values to avoid `tl.sort` corruption
3. Sort ascending via `tl.sort`
4. Extract median value at the middle position
5. Restore `inf`/`-inf` values for median position if needed
6. Find original index using occurrence counting method (handles duplicate values correctly)

### Key Code

```python
@libentry()
@triton.jit
def median_kernel_float(
    inp_ptr,
    out_val_ptr,
    out_idx_ptr,
    N: tl.constexpr,
    PADDED_N: tl.constexpr,
    MED_POS: tl.constexpr,
):
    pid = tle.program_id(0)
    inp_row_ptr = inp_ptr + pid * N
    out_val_ptr_pid = out_val_ptr + pid
    out_idx_ptr_pid = out_idx_ptr + pid

    cols = tl.arange(0, PADDED_N)
    mask = cols < N

    # Use max float value (not inf, which causes tl.sort to produce nan in Triton 3.1)
    pad_val = tl.full([PADDED_N], dtype=tl.float32, value=3.4028235e38)
    vals = tl.load(inp_row_ptr + cols, mask=mask, other=0.0).to(tl.float32)
    vals = tl.where(mask, vals, pad_val)

    # Replace inf/-inf with large finite values to avoid tl.sort corruption
    has_neg_inf = mask & (vals == float('-inf'))
    has_inf = mask & (vals == float('inf'))
    vals = tl.where(has_neg_inf, -3.4028235e38, vals)
    vals = tl.where(has_inf, 3.4028235e38, vals)

    # Sort ascending
    sorted_vals = tl.sort(vals, dim=0, descending=False)

    # Extract median value using positional mask
    med_val = tl.sum(tl.where(cols == MED_POS, sorted_vals, 0.0))

    # Restore inf/-inf for median value if needed
    count_neg_inf = tl.sum(has_neg_inf.to(tl.int32))
    count_inf = tl.sum(has_inf.to(tl.int32))
    med_val = tl.where(MED_POS < count_neg_inf, float('-inf'), med_val)
    med_val = tl.where(MED_POS >= PADDED_N - count_inf, float('inf'), med_val)

    # Find original index (handles duplicates via occurrence counting)
    search_val = tl.where(med_val == float('-inf'), -3.4028235e38, med_val)
    search_val = tl.where(med_val == float('inf'), 3.4028235e38, search_val)
    matches = mask & (vals == search_val)
    is_less = mask & (vals < search_val)
    count_less = tl.sum(is_less.to(tl.int32))
    occurrence = MED_POS - count_less + 1
    match_cumsum = tl.cumsum(matches.to(tl.int32), axis=0)
    is_target = matches & (match_cumsum == occurrence)
    orig_idx = tl.sum(tl.where(is_target, cols, 0))

    tl.store(out_val_ptr_pid, med_val)
    tl.store(out_idx_ptr_pid, orig_idx.to(tl.int64))
```

### Design Decisions

| Decision | Choice | Reason |
|----------|--------|--------|
| Sorting algorithm | `tl.sort` | Triton native optimization, cross-platform compatible |
| Padding value | `3.4028235e38` (max float) | Avoid `inf` causing `tl.sort` to produce `nan` |
| Inf handling | Replace with large finite values | `tl.sort` corrupts `inf`/`-inf` values |
| Index finding | Occurrence counting | Correctly handles duplicates using `tl.cumsum` |
| Dtype handling | Separate float/int kernels | Prevent precision loss for large integers |

### Supported Data Types

| dtype | Status | Precision Requirement |
|-------|--------|----------------------|
| `torch.float32` | Pass | `rtol=1e-5, atol=1.3e-6` |
| `torch.float16` | Pass | `rtol=1e-4, atol=1e-3` |
| `torch.bfloat16` | Pass | `rtol=1e-4, atol=0.016` |
| `torch.int32` | Pass | Exact match |
| `torch.int64` | Pass | Exact match |

### Test Results (Cross-Platform)

| Platform | Tests | Result |
|----------|-------|--------|
| NVIDIA RTX 4090 | 177 passed, 78 skipped | All pass |
| Iluvatar BI-V150 | 177 passed, 78 skipped | All pass |

All tests pass including:
- Precision verification against PyTorch reference
- Index verification (returned index points to correct median value)
- Edge cases: NaN propagation, Inf handling, empty tensors

### Benchmark Results (NVIDIA RTX 4090)

**Method**: `triton.testing.do_bench` - PyTorch baseline measured before `flag_gems.enable()`

#### float32

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0072 | 0.0061 | 1.167x |
| (256, 256) | 0.0082 | 0.0082 | 1.000x |
| (1024, 1024) | 0.0481 | 0.0430 | 1.119x |
| (1024, 1) | 0.0061 | 0.0041 | 1.500x |
| (1024, 512) | 0.0225 | 0.0215 | 1.048x |

#### float16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0061 | 0.0082 | 0.750x |
| (256, 256) | 0.0072 | 0.0092 | 0.778x |
| (1024, 1024) | 0.0410 | 0.0502 | 0.816x |
| (1024, 1) | 0.0061 | 0.0041 | 1.500x |
| (1024, 512) | 0.0195 | 0.0225 | 0.864x |

#### bfloat16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0072 | 0.0072 | 1.000x |
| (256, 256) | 0.0082 | 0.0082 | 1.000x |
| (1024, 1024) | 0.0481 | 0.0492 | 0.979x |
| (1024, 1) | 0.0061 | 0.0041 | 1.500x |
| (1024, 512) | 0.0215 | 0.0225 | 0.955x |

### Benchmark Results (Iluvatar BI-V150)

**Method**: FlagGems built-in benchmark framework

#### float32

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0189 | 0.0192 | 0.981x |
| (256, 256) | 0.0251 | 0.0250 | 1.007x |
| (1024, 1024) | 0.2170 | 0.2168 | 1.001x |
| (4096, 4096) | 1.5893 | 1.5883 | 1.001x |
| (1024, 65536) | 4.6402 | 4.6393 | 1.000x |
| (1024, 1) | 0.0148 | 0.0152 | 0.971x |
| (1024, 512) | 0.1138 | 0.1138 | 1.000x |
| (16, 131072) | 0.5575 | 0.5567 | 1.001x |
| (8, 262144) | 1.1988 | 1.1982 | 1.000x |

#### float16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0185 | 0.0197 | 0.940x |
| (256, 256) | 0.0233 | 0.0230 | 1.013x |
| (1024, 1024) | 0.1895 | 0.1898 | 0.998x |
| (4096, 4096) | 1.2633 | 1.2626 | 1.001x |
| (1024, 65536) | 3.4752 | 3.4844 | 0.997x |
| (1024, 1) | 0.0151 | 0.0146 | 1.030x |
| (1024, 512) | 0.0981 | 0.0985 | 0.996x |
| (16, 131072) | 0.4747 | 0.4744 | 1.001x |
| (8, 262144) | 0.9721 | 0.9728 | 0.999x |

#### bfloat16

| Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------------|---------------|---------|
| (64, 64) | 0.0198 | 0.0191 | 1.038x |
| (256, 256) | 0.0258 | 0.0259 | 0.997x |
| (1024, 1024) | 0.2131 | 0.2139 | 0.996x |
| (4096, 4096) | 1.4195 | 1.4204 | 0.999x |
| (1024, 65536) | 3.8037 | 3.8067 | 0.999x |
| (1024, 1) | 0.0148 | 0.0145 | 1.025x |
| (1024, 512) | 0.1106 | 0.1107 | 0.999x |
| (16, 131072) | 0.5016 | 0.5004 | 1.002x |
| (8, 262144) | 0.9388 | 0.9394 | 0.999x |

### Performance Summary (Cross-Platform)

| Platform | dtype | Min Speedup | Avg Speedup | Max Speedup | Meets >= 0.9x |
|----------|-------|-------------|-------------|-------------|---------------|
| RTX 4090 (N<=2048) | float32 | 1.000x | 1.167x | 1.500x | Yes |
| RTX 4090 (N<=2048) | float16 | 0.750x | 0.942x | 1.500x | Mostly |
| RTX 4090 (N<=2048) | bfloat16 | 0.955x | 1.107x | 1.500x | Yes |
| BI-V150 (all N) | float32 | 0.971x | 0.996x | 1.007x | Yes |
| BI-V150 (all N) | float16 | 0.940x | 0.998x | 1.030x | Yes |
| BI-V150 (all N) | bfloat16 | 0.996x | 1.001x | 1.038x | Yes |

### Performance Analysis

**Key findings**:
- **BI-V150**: Achieves ~1.0x speedup across ALL shapes and dtypes, meeting the >= 0.9x requirement
- **RTX 4090 (N <= 2048)**: Triton kernel achieves ~1.0x-1.5x speedup for small/medium shapes
- **RTX 4090 (N > 2048)**: Fallback path is slow because `torch.sort` is also overridden by flag_gems's Triton sort, which is slower than PyTorch's native CUDA sort on NVIDIA

**Root cause for 4090 large shapes**: When `flag_gems.use_gems()` is active, ALL operators are overridden including `torch.sort`. Our fallback calls `torch.sort` for large N, which dispatches to flag_gems's Triton sort kernel. On NVIDIA, this Triton sort is ~10x slower than PyTorch's native CUDA sort. On BI-V150, the Triton sort is competitive (~1.0x), so this is not an issue.

### Code Style
- Python follows PEP 8 specification
- Passes `black`, `isort`, `flake8` pre-commit checks
- Clear variable naming and complete function documentation

### Interface Adaptation
- Fully aligned with PyTorch `torch.median` interface
- Supports `dim` parameter
- Supports `keepdim` parameter
- Returns `(values, indices)` tuple

### Registration
```python
# src/flag_gems/__init__.py
("median.dim", median),

# src/flag_gems/ops/__init__.py
from flag_gems.ops.median import median
```

### Hardware Adaptation

| Platform | Status | Compiler |
|----------|--------|----------|
| NVIDIA RTX 4090 | Fully compatible | Standard Triton 3.1.0 |
| Iluvatar BI-V150 | Fully compatible | FlagTree 0.5.1+iluvatar3.1 |

### Standard Triton APIs

Implementation uses only standard Triton APIs with no hardware-specific code:

| API | Purpose | Cross-Platform |
|-----|---------|----------------|
| `tl.sort` | Sorting | Yes |
| `tl.where` | Conditional selection | Yes |
| `tl.sum` | Reduction | Yes |
| `tl.cumsum` | Prefix sum | Yes |
| `tl.arange` | Index generation | Yes |
| `tl.load/tl.store` | Memory access | Yes |
| `tle.program_id` | Program identification | Yes |

### No NVIDIA-Specific Code
- No `__shfl_sync` or warp-level primitives
- No PTX assembly
- No CUDA-specific intrinsics

### GPU Utilization Data (Iluvatar BI-V150)

**GPU Specifications**:
- Model: Iluvatar BI-V150
- Memory: 32768 MiB HBM
- SM Clock: 1500 MHz (idle) / 1600 MHz (boost)
- Memory Clock: 1600 MHz

**Bandwidth Estimation** (shape: 1024, 65536, float32):
- Input: 1024 x 65536 x 4 bytes = 256 MiB
- Time: ~4.64 ms
- Bandwidth: ~55 GiB/s
- Theoretical Peak: ~512 GiB/s
- Bandwidth Utilization: ~10.7%

#### 1. Input Scale Coverage

| Scale | Shape | Status |
|-------|-------|--------|
| Small | (1,), (2,), (3,), (4,), (7,), (8,) | Covered |
| Medium | (64,), (256, 128), (32, 64) | Covered |
| Large | (1024, 1024), (4096,) | Covered |
| Empty | (0,), (0, 5), (3, 0, 4), (2, 5, 0) | Covered |

#### 2. Dimension Coverage

| Dimension | Test Shape | Status |
|-----------|-----------|--------|
| 1D | (64,), (4096,) | Covered |
| 2D | (2, 32), (256, 128), (1024, 1024) | Covered |
| 3D | (2, 3, 4) | Covered |
| 4D | (2, 3, 4, 5) | Covered |
| 5D | (2, 3, 4, 5, 6) | Covered |

#### 3. Parameter Mode Coverage

| Parameter | Test Values | Status |
|-----------|-------------|--------|
| dim | 0, 1, -1, -2, None | Covered |
| keepdim | True, False | Covered |

#### 4. Data Type Coverage

| dtype | Status |
|-------|--------|
| torch.float32 | Covered |
| torch.float16 | Covered |
| torch.bfloat16 | Covered |
| torch.int32 | Covered |
| torch.int64 | Covered |

#### 5. Edge Case Coverage

| Case | Status |
|------|--------|
| Empty tensor | PASS |
| Single element | PASS |
| All same values | PASS |
| Duplicate values | PASS |
| Non-contiguous tensor (transpose) | PASS |
| NaN propagation | PASS |
| Inf handling | PASS |
| -Inf handling | PASS |
| All NaN | PASS |
| Negative values | PASS |
| Large range values | PASS |
| Two elements (even) | PASS |
| Four elements (even) | PASS |
| Int32 boundary values | PASS |

### Test Function List (21 functions)

1. `test_median_float` - Floating-point type comprehensive test
2. `test_median_int` - Integer type comprehensive test
3. `test_median_small_shapes` - Small tensor shape test
4. `test_median_duplicates` - Duplicate values test
5. `test_median_all_same` - All same values test
6. `test_median_non_contiguous` - Non-contiguous tensor test
7. `test_median_single_element` - Single element test
8. `test_median_high_dim` - High-dimensional tensor test (3D-5D)
9. `test_median_with_nan` - NaN value test
10. `test_median_with_inf` - Inf value test
11. `test_median_with_neg_inf` - -Inf value test
12. `test_median_all_nan` - All NaN test
13. `test_median_negative_values` - Negative values test
14. `test_median_large_range` - Large range values test
15. `test_median_two_elements` - Two elements (even) test
16. `test_median_four_elements` - Four elements (even) test
17. `test_median_half_precision` - Half-precision type test
18. `test_median_int64` - int64 test
19. `test_median_int32_boundary` - int32 boundary values test
20. `test_median_keepdim_true` - keepdim=True test
21. `test_median_keepdim_false` - keepdim=False test

### Code Statistics
- `src/flag_gems/ops/median.py`: ~300 lines
- `tests/test_median.py`: ~450 lines
- `benchmark/test_median.py`: ~33 lines

### Documentation
- Complete docstring for each kernel function
- Detailed comments for key algorithm steps
- Clear variable naming
- Explanations for complex logic (max float padding, Inf handling, occurrence counting)

### Code Structure
```python
# Supported dtype definitions
_SUPPORTED_FLOAT_DTYPES = (torch.float16, torch.bfloat16, torch.float32)
_SUPPORTED_INT_DTYPES = (torch.int32, torch.int64)

# Floating-point kernel
@libentry()
@triton.jit
def median_kernel_float(...)

# Integer kernel
@libentry()
@triton.jit
def median_kernel_int(...)

# Main function
def median(inp, dim=None, keepdim=False)
```

## File List

| File | Type | Description |
|------|------|-------------|
| `src/flag_gems/ops/median.py` | New | Core operator implementation |
| `tests/test_median.py` | New | Unit tests (21 test functions) |
| `benchmark/test_median.py` | New | Performance benchmark |
| `src/flag_gems/__init__.py` | Modified | Register median.dim operator |
| `src/flag_gems/ops/__init__.py` | Modified | Import median module |

### NVIDIA RTX 4090
- GPU: NVIDIA GeForce RTX 4090 (24564 MiB GDDR6X)
- Driver: 580.126.09
- PyTorch: 2.6.0+cu124
- Triton: 3.1.0
- Python: 3.12
- OS: Linux 6.8.0-107-generic

### Iluvatar BI-V150
- GPU: Iluvatar BI-V150 x 2 (32GB HBM)
- SDK: Iluvatar CoreX SDK 4.4.0.rc.11.20251201
- CUDA: 10.2 (compatible mode)
- PyTorch: 2.7.1+corex.4.4.0
- Triton: 3.1.0+corex.4.4.0 (FlagTree 0.5.1+iluvatar3.1)
- OS: Linux 5.4.0-148-generic

## Running Tests

```bash
# Set environment
export CUDA_VISIBLE_DEVICES=0

# Run unit tests
pytest tests/test_median.py -v --timeout=120 -k "not shape2"

# Run benchmark
python3 -c "
import torch
import triton
import flag_gems

flag_gems.enable()
inp = torch.randn(1024, 1024, dtype=torch.float32, device='cuda')
fn = lambda: torch.median(inp, dim=-1)
latency = triton.testing.do_bench(fn, warmup=10, rep=100, return_mode='median')
print(f'FlagGems: {latency:.4f}ms')
"
```

## Reference Implementation

- Reduction operator reference: `src/flag_gems/ops/argmax.py`, `src/flag_gems/ops/sort.py`
- Test reference: `tests/test_sort.py`, `tests/test_argmax.py`
- Benchmark framework: `benchmark/base.py`

### CODE REVIEW

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
